modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/service/ClientServiceInvokeRequest.java

```diff
@@ -0,0 +1,324 @@
+ */
+
+package org.apache.ignite.internal.processors.platform.client.service;
+
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.ignite.IgniteException;
+import org.apache.ignite.IgniteServices;
+import org.apache.ignite.internal.binary.BinaryContext;
+import org.apache.ignite.internal.binary.BinaryRawReaderEx;
+import org.apache.ignite.internal.binary.BinaryReaderExImpl;
+import org.apache.ignite.internal.cluster.ClusterGroupAdapter;
+import org.apache.ignite.internal.processors.cache.binary.CacheObjectBinaryProcessorImpl;
+import org.apache.ignite.internal.processors.platform.PlatformNativeException;
+import org.apache.ignite.internal.processors.platform.client.ClientConnectionContext;
+import org.apache.ignite.internal.processors.platform.client.ClientObjectResponse;
+import org.apache.ignite.internal.processors.platform.client.ClientRequest;
+import org.apache.ignite.internal.processors.platform.client.ClientResponse;
+import org.apache.ignite.internal.processors.platform.services.PlatformService;
+import org.apache.ignite.internal.processors.platform.services.PlatformServices;
+import org.apache.ignite.internal.processors.service.GridServiceProxy;
+import org.apache.ignite.internal.util.typedef.F;
+import org.apache.ignite.internal.util.typedef.internal.S;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.services.Service;
+import org.apache.ignite.services.ServiceDescriptor;
+
+/**
+ * Request to invoke service method.
+ */
+public class ClientServiceInvokeRequest extends ClientRequest {
+    /** Flag keep binary mask. */
+    private static final byte FLAG_KEEP_BINARY_MASK = 0x01;
+
+    /** Flag "has parameter types", indicates that method arguments prefixed by typeId to help to resolve methods. */
+    private static final byte FLAG_PARAMETER_TYPES_MASK = 0x02;
+
+    /** Methods cache. */
+    private static final Map<MethodDescriptor, Method> methodsCache = new ConcurrentHashMap<>();
+
+    /** Service name. */
+    private final String name;
+
+    /** Flags. */
+    private final byte flags;
+
+    /** Timeout. */
+    private final long timeout;
+
+    /** Nodes. */
+    private final Set<UUID> nodeIds;
+
+    /** Method name. */
+    private final String methodName;
+
+    /** Method parameter type IDs. */
+    private final int[] paramTypeIds;
+
+    /** Service arguments. */
+    private final Object[] args;
+
+    /** Objects reader. */
+    private final BinaryRawReaderEx reader;
+
+    /**
+     * Constructor.
+     *
+     * @param reader Reader.
+     */
+    public ClientServiceInvokeRequest(BinaryReaderExImpl reader) {
+        super(reader);
+
+        name = reader.readString();
+
+        flags = reader.readByte();
+
+        timeout = reader.readLong();
+
+        int cnt = reader.readInt();
+
+        nodeIds = U.newHashSet(cnt);
```

Comment:
Let's change `nodeIds` to `Collection` - looks like there is no need for a hash set.

Reply:
Fixed